src/Npgsql/NpgsqlDiagnosticListenerExtensions.cs

```diff
@@ -0,0 +1,332 @@
+        public const string NpgsqlAfterExecuteCommand = NpgsqlClientPrefix + nameof(WriteCommandAfter);
+        public const string NpgsqlErrorExecuteCommand = NpgsqlClientPrefix + nameof(WriteCommandError);
+
+        public const string NpgsqlBeforeOpenConnection = NpgsqlClientPrefix + nameof(WriteConnectionOpenBefore);
+        public const string NpgsqlAfterOpenConnection = NpgsqlClientPrefix + nameof(WriteConnectionOpenAfter);
+        public const string NpgsqlErrorOpenConnection = NpgsqlClientPrefix + nameof(WriteConnectionOpenError);
+
+        public const string NpgsqlBeforeCloseConnection = NpgsqlClientPrefix + nameof(WriteConnectionCloseBefore);
+        public const string NpgsqlAfterCloseConnection = NpgsqlClientPrefix + nameof(WriteConnectionCloseAfter);
+        public const string NpgsqlErrorCloseConnection = NpgsqlClientPrefix + nameof(WriteConnectionCloseError);
+
+        public const string NpgsqlBeforeCommitTransaction = NpgsqlClientPrefix + nameof(WriteTransactionCommitBefore);
+        public const string NpgsqlAfterCommitTransaction = NpgsqlClientPrefix + nameof(WriteTransactionCommitAfter);
+        public const string NpgsqlErrorCommitTransaction = NpgsqlClientPrefix + nameof(WriteTransactionCommitError);
+
+        public const string NpgsqlBeforeRollbackTransaction = NpgsqlClientPrefix + nameof(WriteTransactionRollbackBefore);
+        public const string NpgsqlAfterRollbackTransaction = NpgsqlClientPrefix + nameof(WriteTransactionRollbackAfter);
+        public const string NpgsqlErrorRollbackTransaction = NpgsqlClientPrefix + nameof(WriteTransactionRollbackError);
+
+        public static Guid WriteCommandBefore(this DiagnosticListener @this, NpgsqlCommand command, [CallerMemberName] string operation = "")
+        {
+            if (!@this.IsEnabled(NpgsqlBeforeExecuteCommand))
+            {
+                return Guid.Empty;
+            }
+
+            var operationId = Guid.NewGuid();
+
+            @this.Write(
+                NpgsqlBeforeExecuteCommand,
+                new
+                {
+                    OperationId = operationId,
+                    Operation = operation,
+                    ConnectionId = command.Connection?.Connector?.Id,
+                    Command = command
+                });
+
+            return operationId;
+        }
+
+        public static void WriteCommandAfter(this DiagnosticListener @this, Guid operationId, NpgsqlCommand command, [CallerMemberName] string operation = "")
+        {
+            if (@this.IsEnabled(NpgsqlAfterExecuteCommand))
+            {
+                @this.Write(
+                    NpgsqlAfterExecuteCommand,
+                    new
+                    {
+                        OperationId = operationId,
+                        Operation = operation,
+                        ConnectionId = command.Connection?.Connector?.Id,
+                        Command = command,
+                        Timestamp = Stopwatch.GetTimestamp()
```

Comment:
I'm not sure it's the job of the event producer to get the timestamp - seems to me like that's supposed to be done at the subscriber (receiver side).... As I wrote in the general review comments, the idea is also to eliminate the anonymous type allocation here by simply passing the NpgsqlCommand instance.